src/lib.rs

```diff
@@ -516,6 +516,9 @@ impl<T> Async<T> {
     /// # std::io::Result::Ok(()) });
     /// ```
     pub async fn read_with<R>(&self, op: impl FnMut(&T) -> io::Result<R>) -> io::Result<R> {
+        if self.source.readers_registered() {
```

Comment:
Should this be `if !self.source.readers_registered()`?

So if there are readers waiting on this I/O object, then we can optimistically try a read operation. But if there are readers waiting on it, then we should just skip it and go register ourselves.

Reply:
The line below is not the optimistic read. If we have registered readers, we wait for the source to be readable (thus registering ourselves) before attempting the read for the first time

Reply:
Ah you're right. Sorry, I think the `optimistic()` threw me off here. 😅

Reply:
Could you perhaps move that check into the loop, so something like:

```rust
// If there are no blocked readers, attempt the read operation.
if !self.source.readers_registered() {
    match op(self.get_ref()) {
        // ...
    }
}
// Wait until the I/O handle becomes readable.
optimistic(self.readable()).await?;
```

Reply:
Done